core/reactor_threaded.c

```diff
@@ -1014,18 +1014,227 @@ int worker_thread_count = 0;
+
+    if (_lf_worker_should_stop()) {
+        return true;
+    }
+
+    _lf_logical_tag_completed = true;
+
+    // Advance time.
+    // _lf_next() may block waiting for real time to pass or events to appear.
+    // to appear on the event queue. Note that we already
+    // hold the mutex lock.
+    tracepoint_worker_advancing_time_starts(worker_number);
+    _lf_next();
+    tracepoint_worker_advancing_time_ends(worker_number);
+    _lf_advancing_time = false;
+    DEBUG_PRINT("Worker %d: Done waiting for _lf_next().", worker_number);
+    return false;
+}
+
+/**
+ * Advance tag if there are no reactions in the reaction queue or in progress. If
+ * there are such reactions or if another thread is already advancing time, wait
+ * until something on the reaction queue is changed.
+ * 
+ * @return should_exit True if the worker thread should exit. False otherwise.
+ */
+bool _lf_worker_try_advance_tag_or_wait(int worker_number) {
+    if (pqueue_size(reaction_q) == 0
+            && pqueue_size(executing_q) == 0) {
+            // Nothing more happening at this logical time.
+        if (!_lf_advancing_time) {
+            // This thread will take charge of advancing time.
+            if (_lf_worker_advance_tag(worker_number)) {
+                return true;
+            } else {
+                return false;
+            }
+        } else if (compare_tags(current_tag, stop_tag) >= 0) {
+            // At the stop tag so we can exit the worker thread.
+            return true;
+        }
+    }
+    // We have not returned, so logical time is not complete or some other worker thread
+    // is advancing time.
+
+    // Wait for something to change (either a stop request or
+    // something went on the reaction queue).
+    DEBUG_PRINT("Worker %d: Waiting for items on the reaction queue.", worker_number);
+    tracepoint_worker_wait_starts(worker_number);
+    // NOTE: Could use a timedwait here to ensure that the worker thread
+    // wakes up periodically. But this appears to be unnecessary. When there
+    // is a ready reaction on the reaction queue, there will be a notification
+    // and notification occurs while holding the mutex lock so it should not be missed.
+    // Nevertheless, we keep the commented out code for a timedwait in case we later
+    // want to put this back in:
+    //
+    // struct timespec physical_time;
+    // lf_clock_gettime(CLOCK_REALTIME, &physical_time);
+    // physical_time.tv_nsec += MAX_STALL_INTERVAL;
+    // lf_cond_wait(&reaction_q_changed, &mutex, &physical_time);
+    lf_cond_wait(&reaction_q_changed, &mutex);
+    tracepoint_worker_wait_ends(worker_number);
+    DEBUG_PRINT("Worker %d: Done waiting.", worker_number);
+
+    return false;
+}
+
+/**
+ * Handle deadline violation for 'reaction'.
+ * 
+ * @return true if a deadline violation occurred. false otherwise.
+ */
+bool _lf_worker_handle_deadline_violation_for_reaction(int worker_number, reaction_t* reaction) {
```

Comment:
I'm guess this function must be called when the mutex lock is NOT held. This should be documented.